Improve Text Extraction

pubpub-zz opened the pull request:
fix at least #880

line feeds are applied as stated in the ref

TODO : for segmented text, horizontal translation should be analysed to apply some space or delete some characters (to be analysed in crazyones.pdf)
TODO : font conversion (to be analysed in crazyones.pdf)

pubpub-zz commented:
@MartinThoma 
 For review only, still some works before merging into 2.0

pubpub-zz commented:
deep changes in the text extractions. still some analysis with some fonts(truetype,type 3)

Looking for so beta testers.

MartinThoma commented:
Uuuh, nice! I'll let it run with https://github.com/py-pdf/benchmarks this evening if I have some time :-)

MartinThoma commented on PyPDF2/_page.py:
```diff
@@ -1061,48 +1064,135 @@ def extract_text(self, Tj_sep: str = "", TJ_sep: str = "") -> str:
+                        else:
+                            try:
+                                encoding[x] = adobe_glyphs[o]
+                            except:
```
```suggestion
                            except Exception:
```

MartinThoma commented on PyPDF2/generic.py:
```diff
@@ -1702,7 +1708,17 @@ def createStringObject(
+        for x in string:
+            try:
+                out += forceEncoding[x]
+            except:
```
```suggestion
            except Exception:
```

MartinThoma commented on PyPDF2/generic.py:
```diff
@@ -2015,14 +2031,68 @@ def decode_pdfdocencoding(byte_array: bytes) -> str:
+    for x in range(256):
+        try:
+            lst+=(bytes((x,)).decode(enc),)
+        except:
```
```suggestion
        except Exception:
```

MartinThoma commented on PyPDF2/_page.py:
```diff
@@ -1061,48 +1064,135 @@ def extract_text(self, Tj_sep: str = "", TJ_sep: str = "") -> str:
+            processRg = False
+            processChar = False
+            encoding = []
+            fontType = pdf.getPage(0)["/Resources"]["/Font"][font_name]["/Subtype"]
```
```suggestion
            fontType = pdf.pages[0]["/Resources"]["/Font"][font_name]["/Subtype"]
```

MartinThoma commented on PyPDF2/_page.py:
```diff
@@ -1061,48 +1064,135 @@ def extract_text(self, Tj_sep: str = "", TJ_sep: str = "") -> str:
+            processChar = False
+            encoding = []
+            fontType = pdf.getPage(0)["/Resources"]["/Font"][font_name]["/Subtype"]
+            if "/Encoding" in pdf.getPage(0)["/Resources"]["/Font"][font_name]:
```
```suggestion
            if "/Encoding" in pdf.pages[0]["/Resources"]["/Font"][font_name]:
```

MartinThoma commented on PyPDF2/_page.py:
```diff
@@ -1061,48 +1064,135 @@ def extract_text(self, Tj_sep: str = "", TJ_sep: str = "") -> str:
+            encoding = []
+            fontType = pdf.getPage(0)["/Resources"]["/Font"][font_name]["/Subtype"]
+            if "/Encoding" in pdf.getPage(0)["/Resources"]["/Font"][font_name]:
+                enc= pdf.getPage(0)["/Resources"]["/Font"][font_name]["/Encoding"].getObject()
```
```suggestion
                enc= pdf.pages[0]["/Resources"]["/Font"][font_name]["/Encoding"].get_object()
```

MartinThoma commented on PyPDF2/_page.py:
```diff
@@ -1061,48 +1064,135 @@ def extract_text(self, Tj_sep: str = "", TJ_sep: str = "") -> str:
+                            except Exception:
+                                encoding[x] = o
+                            x += 1
+            if "/ToUnicode" in pdf.getPage(0)["/Resources"]["/Font"][font_name]:
```
```suggestion
            if "/ToUnicode" in pdf.pages[0]["/Resources"]["/Font"][font_name]:
```

MartinThoma commented on PyPDF2/_page.py:
```diff
@@ -1061,48 +1064,135 @@ def extract_text(self, Tj_sep: str = "", TJ_sep: str = "") -> str:
+                                encoding[x] = o
+                            x += 1
+            if "/ToUnicode" in pdf.getPage(0)["/Resources"]["/Font"][font_name]:
+                cm = pdf.getPage(0)["/Resources"]["/Font"][font_name]["/ToUnicode"].getData().decode('utf-8')
```
```suggestion
                cm = pdf.pages[0]["/Resources"]["/Font"][font_name]["/ToUnicode"].getData().decode('utf-8')
```

MartinThoma commented:
I fixed the merge conflicts due to the recent PEP8 renamings + I adjusted the PR to the new syntax. There are still mypy issues though.

MartinThoma commented:
Running the benchmark, I get:

```
  File "/home/moose/Github/py-pdf/PyPDF2/PyPDF2/_page.py", line 1141, in extract_text
    ft,cmap,cmap2 = buildCharMap(self.pdf,operands[0])
  File "/home/moose/Github/py-pdf/PyPDF2/PyPDF2/_page.py", line 1073, in buildCharMap
    fontType = pdf.pages[0]["/Resources"]["/Font"][font_name]["/Subtype"]
  File "/home/moose/Github/py-pdf/PyPDF2/PyPDF2/generic.py", line 519, in __getitem__
    return dict.__getitem__(self, key).get_object()
KeyError: '/R167'
```

pubpub-zz commented:
@MartinThoma 
> Running the benchmark, I get:
> 
> ```
>   File "/home/moose/Github/py-pdf/PyPDF2/PyPDF2/_page.py", line 1141, in extract_text
>     ft,cmap,cmap2 = buildCharMap(self.pdf,operands[0])
>   File "/home/moose/Github/py-pdf/PyPDF2/PyPDF2/_page.py", line 1073, in buildCharMap
>     fontType = pdf.pages[0]["/Resources"]["/Font"][font_name]["/Subtype"]
>   File "/home/moose/Github/py-pdf/PyPDF2/PyPDF2/generic.py", line 519, in __getitem__
>     return dict.__getitem__(self, key).get_object()
> KeyError: '/R167'
> ```

Thanks for the test and the report. I've fixed it; also I've introduced a test using testbench dataset. using urllib.request to get them. It's increasing the test duration. I've set a way to bypass it.

I think that this pull now should be proposed for testing 

PS : I merged with the new API and 2 functions are missing to prevent compatibility loss : getPage and getObject

MartinThoma commented:
>     def read_from_stream(stream: StreamType, pdf: Any, forcedEncoding: Union[None,str,List[str],dict[int,str]] = None) -> "ArrayObject":  # PdfReader
> E   TypeError: 'type' object is not subscriptable

You need to change `dict[int,str]` to `Dict[int,str]` (capital `D`). The syntax you've used was only introduced in Python 3.9 or 3.10, I think

MartinThoma commented:
A general question: If I see such things, should I directly adjust it in your PR? (I know some people love it and others hate it)

pubpub-zz commented:
Thanks, for the advice. I've tried to improve myself.
I successed to fix it 😊
During your review, I would like to have your opinion : the solution I've used to prevent mypy errors with cast makes the code not so easy to read for me: can you give me some advice ?

MartinThoma commented:
Glancing at the first result of [the benchmark](https://github.com/py-pdf/benchmarks) the results look amazing :heart_eyes: 

![image](https://user-images.githubusercontent.com/1658117/170821756-e535b123-238d-4e15-9bb0-850705c8e954.png)

It will take a while until the benchmark completed. I'll share the results later :-)

MartinThoma commented:
@pubpub-zz Could you please have a look at https://arxiv.org/pdf/1601.03642.pdf ? The text extraction now completely fails on this one. It only extracts Asian characters, although the document doesn't have a single Asian character:

![image](https://user-images.githubusercontent.com/1658117/170822410-9a0f6798-fbad-4734-8b09-c6d8055c2f71.png)

MartinThoma commented:
@pubpub-zz I have the impression you're trying too much at one time here. For example, I can identify at least two different improvements in this PR:

* Spaces (spaceScale)
* CharMap support (buildCharMap)

I guess there are even more.

The spaces improvement seems to add already quite a bit of value. It also seems ready.

Although the CharMap support is the bigger topic, it's also one that is not quite ready yet (due to the Chinese-character issue mentioned above).

I suggest to do those step-by-step. I admire that you're having so many ideas and that you're able to implement them! However, for me as a maintainer it's easier to review smaller chunks. For you as a contributor it might also be more rewarding, because your changes ultimately get merged faster

MartinThoma commented:
> the solution I've used to prevent mypy errors with cast makes the code not so easy to read for me

I'll have a look :-)

MartinThoma commented on PyPDF2/_page.py:
```diff
@@ -1069,48 +1075,188 @@ def extract_text(self, Tj_sep: str = "", TJ_sep: str = "") -> str:
+            processRg : bool = False
+            processChar : bool = False
+            encoding : List[str] = []
+            fontType :str = cast(NameObject, cast(DictionaryObject,
+                                 cast(DictionaryObject,
+                                      cast(DictionaryObject, self["/Resources"])["/Font"]
+                                 )[font_name]
+                            )["/Subtype"])
```
Cast is telling mypy to ignore what it thinks the type is. Typically you only need one cast per line (if you need any at all)

```suggestion
            fontType :str = cast(NameObject, self["/Resources"]["/Font"][font_name]["/Subtype"])
```

MartinThoma commented on PyPDF2/_page.py:
```diff
@@ -1069,48 +1075,188 @@ def extract_text(self, Tj_sep: str = "", TJ_sep: str = "") -> str:
+        text: str = ""
+        output: str = ""
+        cmaps: Dict[str, Tuple[str, Dict[int,str],Dict[str,str]]]={}
+        if "/Font" in cast(DictionaryObject,self["/Resources"]):
+            for f in cast(DictionaryObject,cast(DictionaryObject,self["/Resources"])["/Font"]):
```
```suggestion
        resources_dict = cast(DictionaryObject,self["/Resources"])
        if "/Font" in resources_dict:
            for f in cast(DictionaryObject, resources_dict["/Font"]):
```

MartinThoma commented on PyPDF2/_page.py:
```diff
@@ -1069,48 +1075,188 @@ def extract_text(self, Tj_sep: str = "", TJ_sep: str = "") -> str:
+
+        # charSize = 1.0
+        # charScale = 0.0
+        spaceScale = 1.0
```
Nitpic: I try to get [PEP8](https://peps.python.org/pep-0008/) compliant code. This makes it easier to read as it's one style.

I would still merge any PR that violates PEP8, but I would later clean those lines up.

MartinThoma commented on PyPDF2/generic.py:
```diff
@@ -200,7 +200,7 @@ def writeToStream(
 
     @staticmethod
-    def read_from_stream(stream: StreamType, pdf: Any) -> "ArrayObject":  # PdfReader
+    def read_from_stream(stream: StreamType, pdf: Any, forcedEncoding: Union[None,str,List[str],Dict[int,str]] = None) -> "ArrayObject":  # PdfReader
```
forcedEncoding seems never to be assigned. I guess that is one part that is still a TODO? What does this do?

MartinThoma commented on PyPDF2/_page.py:
```diff
@@ -1069,48 +1075,185 @@ def extract_text(self, Tj_sep: str = "", TJ_sep: str = "") -> str:
+            processRg : bool = False
+            processChar : bool = False
+            encoding : List[str] = []
+            fontType :str = cast(NameObject, self["/Resources"]["/Font"][font_name]["/Subtype"])
+            if "/Encoding" in cast(DictionaryObject, cast(DictionaryObject, cast(DictionaryObject, self["/Resources"])["/Font"])[font_name]):
+                enc = cast(DictionaryObject,
+                           cast(DictionaryObject,
+                                cast(DictionaryObject,
+                                     cast(DictionaryObject, self["/Resources"])["/Font"]
+                                )[font_name]
+                            )["/Encoding"]
```
```suggestion
            resources_dict = cast(DictionaryObject, self["/Resources"])
            fontType :str = cast(NameObject, resources_dict["/Font"][font_name]["/Subtype"])
            if "/Encoding" in resources_dict["/Font"][font_name]:
                enc = resources_dict["/Font"][font_name]["/Encoding"]
```

MartinThoma commented:
Results of the benchmark run (the numbers are the documents):

1. 89% ➔ 89% -- big improvement!
2. 79% ➔ 85% -- big improvement!
3. 89% -> 71% :cry: -- quite a lot of spacing issues + greek characters (maybe cmap?)
4. 91% -> 93%
5. 95% -> 93% -- Meld fails to show a propper diff
6. 92% -> 93% -- Meld fails to show a propper diff
7. 90% -> 88% -- Meld fails to show a propper diff
8. 89% -> 91% -- Meld fails to show a propper diff
9. 97% -> 95% -- Meld fails to show a propper diff
10. 90% -> 90%
11. 97% -> 95%
12. 91% -> 92% -- Meld fails to show a propper diff
13. 0% -> 0%
14. 94% -> 94% -- Meld fails to show a propper diff

I came a big step closer: It seems as if my diff tool ( https://meldmerge.org/ ) had issues opening the file

MartinThoma commented:
It seems like there are some extracted characters that are not in UTF-8:

![image](https://user-images.githubusercontent.com/1658117/170825292-201db305-5c51-44f3-a9d3-365dbdb2aab7.png)

Hence meld falls back to UTF-16 which is just wrong. I'm uncertain if PyPDF2 can (or should) do anything here.

pubpub-zz commented:
> Regarding the Asian character issue: It seems like there are some extracted characters that are not in UTF-8:
> 
> ![image](https://user-images.githubusercontent.com/1658117/170825292-201db305-5c51-44f3-a9d3-365dbdb2aab7.png)
> 
> Hence meld falls back to UTF-16 which is just wrong. I'm uncertain if PyPDF2 can (or should) do anything here.

@MartinThoma
This is my code:
`
import io,urllib.request,PyPDF2;p = PyPDF2.PdfReader(io.BytesIO(urllib.request.urlopen("https://arxiv.org/pdf/1601.03642.pdf").read()));e=p.pages[0].extract_text();print(e);f=open("e:/extracted.txt","wb");f.write(e.encode("utf-8"));f.close()
`
I'm getting some good results.

pubpub-zz commented:
@MartinThoma ,
Unless I'm wrong your proposal seems to not make mypy happy 😕

MartinThoma commented:
@pubpub-zz You're right, I missed some things: https://github.com/py-pdf/PyPDF2/pull/881#discussion_r884123778

pubpub-zz commented:
@MartinThoma,
My latest status
a) about issues with asian characters : I've not been able to reproduce the issue.
my code in python to get the file:
`import io,urllib.request,PyPDF2;p = PyPDF2.PdfReader(io.BytesIO(urllib.request.urlopen("https://arxiv.org/pdf/2201.00151.pdf").read()));e=p.pages[0].extract_text();print(e);f=open('e:/out.txt','w',encoding='utf-8');f.write(e);f.close()`
the result of comparison between out.txt(left) and copy/paste from adobe reader DC: 
![image](https://user-images.githubusercontent.com/4083478/170863523-e0258f86-0e24-4681-968f-4bcef049a1f0.png)

b) spacing:
I've sligthtly modified the code to allow to pass the space width criteria. I' still do not understand some case where a lower width is sometimes interpreted as a space whereas is some cases not.

c) mypy.
Your proposal seems to have reintroduced some error reporting. Can you check on your side and reajust enventually?

d)benchmark
Can you tell me how are you obtaining the percentage result?

MartinThoma commented:
I've extracted the spacing improvement heuristic to https://github.com/py-pdf/PyPDF2/pull/922 . That should be possible to get merged really soon

MartinThoma commented:
I've just fixed the merge conflict I've introduced

MartinThoma commented:
@pubpub-zz Here is the simplification for the casts: https://github.com/py-pdf/PyPDF2/pull/924/commits/c4887349924127a46852e80397909c9dffa28c04 - essentially just assigning variable names to intermediate results

LucasWolfgang commented:
Hey man, just want to say thank you for you hard work on this one. I got some interest on do this improved myself and end up finding this PR. Really nice work, @pubpub-zz. Are you checking list your past and future updates somewhere? I would like to follow and see if I can help to add html features to it (while avoiding disturb your work)

pubpub-zz commented:
> 

@MartinThoma, your proposal is quite heavy to read. May I propose this solution?
`            font_type: str = self["/Resources"]["/Font"][font_name]["/Subtype"]    # type: ignore`

MartinThoma commented:
That is completely fine for me as well :-)

pubpub-zz commented:
> > 
> 
> @MartinThoma, your proposal is quite heavy to read. May I propose this solution? ` font_type: str = self["/Resources"]["/Font"][font_name]["/Subtype"] # type: ignore`

@LucasWolfgang, thanks for your nice comments. There will be still some improvement to be done. Don't know i will be able to release new changes.
What do you intend to propose (font? style? position?)

LucasWolfgang commented:
@pubpub-zz , I was thinking of implemting the functionalities available in PyMuPDF's [get_text](https://pymupdf.readthedocs.io/en/latest/tutorial.html#extracting-text-and-images). Of course it would be a LOT of work. So I would start fist by just creating the blocks (paragraphs) and continue from there to extract font, style and Images (as Base64 or maybe as references) a PR at a time. About positions, I am still unsure how I would use it besides ordering the blocks.

pubpub-zz commented:
maybe what you could do would be to branch from my dev branch https://github.com/pubpub-zz/PyPDF2/tree/extractText
If you had an as_html boolean parameter to add so if branches to be able keep text and html at the same time. This solution may help you to merge my changes later

MartinThoma commented:
@pubpub-zz #924 got merged :tada: I'd make a minor release (2.1.0) on Monday.

MartinThoma commented:
I've moved the buildCharMap function to its own module: https://github.com/py-pdf/PyPDF2/commit/4baedb2054648b9db70c2fbed82b6902aea52365 - I hope this simplifies future PRs / makes it easier for me to understand what they are actually doing.
